plugins/usbdmx/ShowJockey.cpp

```diff
@@ -0,0 +1,392 @@
+
+      if (retVal) {
+        break;
+      }
+
+      buf += transferred;
+      nLeft -= transferred;
+      nPacket = maxPacketSize;
+      transferredAll += transferred;
+    }
+
+    if (transferredAll == size) {
+      return transferredAll;
+    }
+  }
+
+  return -1;
+}
+
+// SynchronousShowJockey
+// -----------------------------------------------------------------------------
+
+SynchronousShowJockey::SynchronousShowJockey(
+    LibUsbAdaptor *adaptor,
+    libusb_device *usb_device,
+    const std::string &serial)
+    : ShowJockey(adaptor, usb_device, serial) {
+}
+
+bool SynchronousShowJockey::Init() {
+  libusb_device_handle *usb_handle;
+
+  int interface_number;
+  int endpoint;
+  int maxPacketSize;
+  if (!LocateInterface(m_adaptor,
+                       m_usb_device,
+                       &interface_number,
+                       &endpoint,
+                       &maxPacketSize)) {
+    return false;
+  }
+
+  bool ok = m_adaptor->OpenDeviceAndClaimInterface(m_usb_device,
+                                                   interface_number,
+                                                   &usb_handle);
+  if (!ok) {
+    return false;
+  }
+
+  std::auto_ptr<ShowJockeyThreadedSender> sender(
+      new ShowJockeyThreadedSender(m_adaptor,
+                                   m_usb_device,
+                                   usb_handle,
+                                   maxPacketSize,
+                                   endpoint));
+  if (!sender->Start()) {
+    return false;
+  }
+
+  m_sender.reset(sender.release());
+  return true;
+}
+
+bool SynchronousShowJockey::SendDMX(const DmxBuffer &buffer) {
+  return m_sender.get() ? m_sender->SendDMX(buffer) : false;
+}
+
+// ShowJockeyAsyncUsbSender
+// -----------------------------------------------------------------------------
+class ShowJockeyAsyncUsbSender : public AsyncUsbSender {
+ public:
+  ShowJockeyAsyncUsbSender(LibUsbAdaptor *adaptor,
+                           libusb_device *usb_device,
+                           int endpoint,
+                           int maxPacketSizeOut,
+                           libusb_device_handle *handle)
+                           : AsyncUsbSender(adaptor, usb_device),
+                             m_endpoint(endpoint),
+                             m_maxPacketSizeOut(maxPacketSizeOut) {
+    m_usb_handle = handle;
+    m_tx_frame = NULL;
+  }
+
+  ~ShowJockeyAsyncUsbSender() {
+    CancelTransfer();
+    delete[] m_tx_frame;
+  }
+
+  libusb_device_handle* SetupHandle() {
+    return m_usb_handle;
+  }
+
+  bool PerformTransfer(const DmxBuffer &buffer) {
+    uint16_t numberOfSequence = DMX_MAX_SLOT_NUMBER / (m_maxPacketSizeOut-2);
+    int finalSize = DMX_MAX_SLOT_NUMBER + 2*numberOfSequence;
+    if (m_tx_frame == NULL) {
+      m_tx_frame = new uint8_t[finalSize]();
+    }
+
+    uint8_t *currentBuffer = new uint8_t[DMX_MAX_SLOT_NUMBER]();
+    memcpy(currentBuffer, buffer.GetRaw(), DMX_MAX_SLOT_NUMBER);
```

Comment:
Can you not use http://docs.openlighting.org/ola/doc/latest/classola_1_1_dmx_buffer.html#ad78488db3fec3fd5e47803d77ef22e5e ?